misk/src/main/kotlin/misk/hibernate/DbTimestampedEntity.kt

```diff
@@ -0,0 +1,54 @@
+package misk.hibernate
+
+import java.time.Instant
+
+/**
+ * This complements [DbEntity] for entities that keep updated_at and created_at timestamps.
+ *
+ * Misk uses the application's clock time to set [created_at] and [updated_at]. We prefer the
+ * application's (ie. JVM's) clock over the database's clock because only the application's clock
+ * can be faked in tests.
+ *
+ * To use timestamped entities first add two columns in your `CREATE TABLE` statement. You may also
+ * add an optional index on `updated_at`:
+ *
+ * ```
+ * CREATE TABLE movies(
+ *   id bigint(20) NOT NULL PRIMARY KEY AUTO_INCREMENT,
+ *   created_at timestamp(3) NOT NULL DEFAULT NOW(3),
+ *   updated_at timestamp(3) NOT NULL DEFAULT NOW(3) ON UPDATE NOW(3),
+ *   ...
+ *   KEY `idx_updated_at` (`updated_at`),
+ * );
+ * ```
+ *
+ * The above SQL uses `timestamp(3)` and `NOW(3)` to get millisecond precision which interoperates
+ * nicely with Java. Although we declare `DEFAULT NOW(3)` and `ON UPDATE NOW(3)`, these don't apply
+ * to writes from Hibernate because it always provides a value explicitly.
+ *
+ * Next implement this interface and add the following declarations to your entity class:
+ *
+ * ```
+ * @Entity
+ * @Table(name = "movies")
+ * class DbMovie() : DbEntity<DbMovie>, DbTimestampedEntity {
+ *
+ *   ...
+ *
+ *   @Column
+ *   override lateinit var updated_at: Instant
+ *
+ *   @Column
+ *   override lateinit var created_at: Instant
+ *
+ *   ...
+ *
+ * }
+ * ```
+ *
+ * The timestamp listener will automatically populate these values on save and update.
+ */
+interface DbTimestampedEntity {
```

Comment:
do you think there is any value in splitting this interface up into mutable and immutable entities? I always found it unfortunate that an immutable entity was forced to have an `updated_at` field

Reply:
Yeah, in this case I’m not quite ready to add that flexibility. In the spirit of being opinionated I want to have fewer types so our users have fewer decisions.

If you really need it, on immutable entities it’s very easy to set your own created_at date.

Reply:
yea good point!